service_configuration_lib/spark_config.py

```diff
@@ -428,30 +434,104 @@ def compute_executor_instances_k8s(user_spark_opts: Dict[str, str]) -> int:
     return executor_instances
 
 
-def _cap_executor_resources(user_spark_opts: Dict[str, str]) -> Dict[str, str]:
-    executor_memory = user_spark_opts.get('spark.executor.memory', DEFAULT_EXECUTOR_MEMORY)
+def _cap_executor_resources(
+    executor_cores: int,
+    executor_memory: str,
+    memory_mb: int,
+) -> Tuple[int, str]:
+    if memory_mb > MAX_EXECUTOR_MEMORY_ALLOWED * 1024:
+        executor_memory = f'{MAX_EXECUTOR_MEMORY_ALLOWED}g'
+        log.warning(
+            f'Given executor memory is {memory_mb / 1024}g, '
+            f'=> capped to {executor_memory} to better fit on available aws nodes.',
+        )
+    elif memory_mb > RECOMMENDED_EXECUTOR_MEMORY * 1024:
+        log.warning(
+            f'We recommend using setting memory as {RECOMMENDED_EXECUTOR_MEMORY}g '
+            f'and executor cores as 4',
+        )
+
+    if executor_cores > MAX_EXECUTOR_CORES_ALLOWED:
+        log.warning(
+            f'Given executor cores is {executor_cores}, '
+            f'=> capped to {MAX_EXECUTOR_CORES_ALLOWED} to better fit on available aws nodes.',
+        )
+        executor_cores = MAX_EXECUTOR_CORES_ALLOWED
+
+    return executor_cores, executor_memory
+
+
+def _recalculate_executor_resources(
+    user_spark_opts: Dict[str, str],
+    force_spark_resource_configs: bool,
+) -> Dict[str, str]:
     executor_cores = int(user_spark_opts.get('spark.executor.cores', str(DEFAULT_EXECUTOR_CORES)))
+    executor_memory = user_spark_opts.get('spark.executor.memory', f'{DEFAULT_EXECUTOR_MEMORY}g')
+    executor_instances = int(user_spark_opts.get('spark.executor.instances', str(DEFAULT_EXECUTOR_INSTANCES)))
+    task_cpus = int(user_spark_opts.get('spark.task.cpus', str(DEFAULT_TASK_CPUS)))
 
     memory_mb = parse_memory_string(executor_memory)
-    if memory_mb > EXECUTOR_MEMORY_CAP_GB * 1024:
-        executor_memory = f'{EXECUTOR_MEMORY_CAP_GB}g'
+    memory_gb = math.ceil(memory_mb / 1024)
+
+    def _calculate_resources(
+        cpu, memory, instances, task_cpus, target_memory,
+    ) -> Tuple[int, str, int, int]:
+        """
+        Calculate resource needed based on memory size and recommended mem:core ratio (7:1).
+
+        Parameters:
+        memory: integer values in GB.
+
+        Returns:
+        A tuple of (new_cpu, new_memory, new_instances, task_cpus).
+        """
+        new_cpu: int
+        new_memory: int
+        new_instances = (instances * memory) // target_memory
```

Comment:
Adding a comment here would be useful.
This can also reduce the instance count if this condition is met: `executor_cores > MEDIUM_EXECUTOR_CORES`. 

Something to keep in mind when rolling this out. If we start noticing to many job failures or timeouts on #spark-devops-notifications channel, we can change this to previous calculation, `max(instances_from_memory, instances_from_cpu)`.

Reply:
Added an explanation in the function description.